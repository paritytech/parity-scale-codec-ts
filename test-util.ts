/// <reference lib="deno.unstable"/>

import { dirname, join } from "std/path/win32.ts";
import { assertEquals } from "std/testing/asserts.ts";
import { assertSnapshot } from "std/testing/snapshot.ts";
import { Codec } from "./common.ts";

const file = (path: string) => () => fetch(join(dirname(import.meta.url), path)).then((x) => x.text());
export const files = {
  lipsum: file("lipsum.txt"),
  words: file("words.txt"),
  cargoLock: file("Cargo.lock"),
};

export function testCodec<T>(name: string, codec: Codec<T>, values: T[] | Record<string, T>) {
  for (const key in values) {
    const value = values[key as never] as T;
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key;
    Deno.test(`${name} ${label}`, async (t) => {
      const encoded = codec.encode(value);
      const decoded = codec.decode(encoded);
      assertEquals(decoded, value);
      await assertSnapshot(t, encoded, { serializer: serializeU8A });
    });
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n");
}

export function benchCodec<T>(name: string, codec: Codec<T>, value: T) {
  const encoded = codec.encode(value);
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    codec.encode(value);
  });
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    codec.decode(encoded);
  });
}
