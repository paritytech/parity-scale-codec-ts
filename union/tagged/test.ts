import * as $ from "../../mod.ts";
import { testCodec } from "../../test-util.ts";

const $abc = $.taggedUnion(
  "_tag",
  ["A"],
  ["B", ["B", $.str]],
  ["C", ["C", $.tuple($.u32, $.u64)]],
  ["D", ["a", $.u32], ["b", $.u64]],
);

testCodec("union", $abc, [
  { _tag: "A" },
  { _tag: "B", B: "HELLO" },
  { _tag: "C", C: [255, 101010101n] },
  { _tag: "D", a: 101, b: 999n },
]);
