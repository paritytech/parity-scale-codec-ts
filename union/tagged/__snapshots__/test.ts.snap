export const snapshot = {};

snapshot[`union { _tag: "A" } 1`] = `00`;

snapshot[`union { _tag: "B", B: "HELLO" } 1`] = `
01
14
48
45
4c
4c
4f
`;

snapshot[`union { _tag: "C", C: [ 255, 101010101n ] } 1`] = `
02
ff
00
00
00
b5
4a
05
06
00
00
00
00
`;

snapshot[`union { _tag: "D", a: 101, b: 999n } 1`] = `
03
65
00
00
00
e7
03
00
00
00
00
00
00
`;
